services/rfq/relayer/inventory/rebalance.go

```diff
@@ -27,183 +27,185 @@ type RebalanceManager interface {
 	}
-	if amount == nil {
-		if span != nil {
-			span.SetAttributes(attribute.Bool("no_rebalance_amount", true))
+
+	// now we have candidate chains, produce the rebalance data for each permutation of the chains
+	rebalanceCandidates := []RebalanceData{}
+	for i := range candidateChains {
+		for j := range candidateChains {
+			if i == j {
+				continue
+			}
+
+			candidate := RebalanceData{
+				OriginMetadata: candidateChains[i],
+				DestMetadata:   candidateChains[j],
+				Method:         method,
+			}
+			rebalanceCandidates = append(rebalanceCandidates, candidate)
 		}
-		return nil, nil
 	}
 
-	rebalance = &RebalanceData{
-		OriginMetadata: originTokenData,
-		DestMetadata:   destTokenData,
-		Amount:         amount,
-		Method:         method,
+	rebalance = getBestRebalance(rebalanceCandidates)
+	if rebalance != nil {
+		rebalance.Amount, err = getRebalanceAmount(ctx, cfg, inv, rebalance)
+		if err != nil {
+			return nil, fmt.Errorf("could not get rebalance amount: %w", err)
+		}
+		if rebalance.Amount == nil {
+			return nil, nil
+		}
 	}
+
 	return rebalance, nil
 }
 
-// getRebalanceMetadatas finds the origin and dest token metadata based on the configured rebalance method.
-//
-//nolint:nestif,cyclop,gocognit
-func getRebalanceMetadatas(cfg relconfig.Config, tokens map[int]map[common.Address]*TokenMetadata, tokenName string, methods []relconfig.RebalanceMethod) (originTokenData, destTokenData *TokenMetadata, method relconfig.RebalanceMethod) {
-	candidates := map[relconfig.RebalanceMethod][2]TokenMetadata{}
-	for _, method := range methods {
-		var originCandidate, destCandidate *TokenMetadata
-		for _, tokenMap := range tokens {
-			for _, tokenData := range tokenMap {
-				if tokenData.Name == tokenName {
-					if !isTokenCompatible(tokenData, method, cfg) {
-						continue
-					}
-
-					// assign origin / dest metadata based on min / max balances
-					if originCandidate == nil || tokenData.Balance.Cmp(originCandidate.Balance) > 0 {
-						originCandidate = tokenData
-					}
-					if destCandidate == nil || tokenData.Balance.Cmp(destCandidate.Balance) < 0 {
-						destCandidate = tokenData
-					}
+// getCandidateChains gets the respective token metadata for each chain that supports the rebalance method.
+func getCandidateChains(cfg relconfig.Config, inv map[int]map[common.Address]*TokenMetadata, method relconfig.RebalanceMethod, tokenName string) (map[int]*TokenMetadata, error) {
+	candidateChains := map[int]*TokenMetadata{}
+	for chainID, chainCfg := range cfg.Chains {
+		var validCandidate bool
+		var candidateMetadata *TokenMetadata
+		for name, tokenCfg := range chainCfg.Tokens {
+			if name != tokenName {
+				continue
+			}
+
+			// check that the token supports given rebalance method
+			if supportsRebalanceMethod(cfg, chainID, tokenCfg.Address, method) {
+				validCandidate = true
+				candidateMetadata = inv[chainID][common.HexToAddress(tokenCfg.Address)]
```

Comment:
Can we just assign to `candidateChains[chainID]` here before breaking and avoid tracking `validCandidate` and `candidateMetadata`?